src/rust/engine/src/nodes.rs

```diff
@@ -1026,37 +1070,67 @@ impl WrappedNode for Task {
     let product = self.product;
     let can_modify_workunit = self.task.can_modify_workunit;
 
-    let result_val =
-      externs::call_function(&externs::val_for(&func.0), &deps).map_err(Failure::from_py_err)?;
-    let mut result_val: Value = result_val.into();
-    let mut result_type = externs::get_type_for(&result_val);
-    if result_type == context.core.types.coroutine {
-      result_val = Self::generate(context.clone(), params, entry, result_val).await?;
-      result_type = externs::get_type_for(&result_val);
-    }
+    let result_val = {
+      // All of the params should have been supplied by coroutines. Ensure that all of the
+      // coroutines succeeded, otherwise error.
+      let mut param_deps: Vec<Value> = vec![];
```

Comment:
I think this code would be cleaner if it iterated over `deps`, checking for an `InvocationResult::Exception` and early-returning if it found one, and then moving on with `deps`; rather than pushing everything from `deps` into `param_deps` in the happy path

Reply:
Ok, I couldn't do that directly, since this code is also unpacking from an `InvocationResult` into a `Value`, but I was able to write it with a `.collect()` instead of iteration, by creating a `From<Result<Vec<Value>, Failure>>` implementation for `InvocationResult`. I believe this simplifies the logic.